ddtrace/tracer.py

```diff
@@ -884,11 +884,17 @@ def set_tags(self, tags):
         """
         self.tags.update(tags)
 
+    def _shutdown_trace(self, *args, **kwargs):
+        raise RuntimeError(
```

Comment:
should we do this as a warning first before we make it a raised exception?

give people 1-2 minor versions to react before raising?

Reply:
What worries me is that if anybody is reusing a tracer they will miss the `shutdown` at exit. This is perhaps not a big deal, but  technically it's still a bug. What we could do is toggle between two different `trace` methods when `shutdown` is called, e.g.
~~~ python
def _trace(...):
    # current trace implementation

def _shutdown_trace(...):
    # restore hooks
    self.trace = self._trace
    self.trace(...)

def shutdown(...):
    ...
    self.trace = self._shutdown_trace
~~~
then we could add the warning, but at this point we might as well allow reusing the tracer, like we do now. The question then becomes whether we _should_ allow it. Based on #2125 the answer is probably no. Does this make sense?

Reply:
I think my objection is more about "is this a breaking change", that we now raise an exception when trying to do this vs before when it was "allowed".

If we want to be safer with the change, then we can do add a warning in this version, and change to an exception in the next.

What do you think?

Reply:
@Kyle-Verhoog and I had a discussion on this as well, these changes sound great and I agree with the approach, but we should avoid any breaking changes before `1.0.0` and instead provide a warning until then.

Reply:
I'm casting my vote for that usage after `shutdown` should be defined to be erroneous and a warning for now and finally an exception in 1.0.

@P403n1x87 raises a good point that we do have a "quirk" now that should be addressed:

```python
tracer.trace("test").finish()
tracer.shutdown()  # atexit handler removed
tracer.trace("test2").finish()  # doesn't get written before proc exits
```

which we either "fix" by defining the functionality of `shutdown` or actually making the above code work by reinstating the atexit handler somehow